piqueserver/config.py

```diff
@@ -0,0 +1,145 @@
+        elif style == JSON_STYLE:
+            self._raw_config.update(json.load(open(config_file)))
+        else:
+            raise ConfigException('Unsupported config file format: {}'.format(style))
+        self.validate_all()
+
+    def load_config_object(self, config):
+        self._raw_config = config
+        self.validate_all()
+
+    def update_config_object(self, config):
+        self._raw_config.update(config)
+        self.validate_all()
+
+    def dump_config(self, out_file, style=DEFAULT_STYLE):
+        if style == TOML_STYLE:
+            toml.dump(self._raw_config, open(out_file, 'w'))
+        elif style == JSON_STYLE:
+            json.dump(self._raw_config, open(out_file, 'w'))
+        else:
+            raise ConfigException('Unsupported config file format: {}'.format(style))
+
+    def get_raw_config(self):
+        return self._raw_config
+
+    def get(self, name, default=None, section=None):
+        if section:
+            if section not in self._raw_config:
+                self._raw_config[section] = {}
+                self._raw_config[section][name] = default
+            return self._raw_config[section][name]
+        else:
+            if name not in self._raw_config:
+                self._raw_config[name] = default
+            return self._raw_config[name]
+
+
+    def set(self, name, value, section=None):
+        if section:
+            if section not in self._raw_config:
+                self._raw_config[section] = {}
+            self._raw_config[section][name] = value
+        else:
+            self._raw_config[name] = value
+
+    def option(self, name, section=None, cast=None, default=None, validate=None):
+        option = Option(self, name, section, default, cast, validate)
+        self.options[(section, name)] = option
+
+        return option
+
+
+
+class Option():
+    '''
+    configuration option object, backed by a configuration store
+    '''
+    def __init__(self, store, name, section, default, cast, validate):
+        self.store = store # ConfigStore object
+        self.name = name
+        self.section = section
+        self.default = default
+        self.cast = cast
+        self.validate_func = validate
+
+        self.validate(self.get())
+
+
+    def validate(self, value):
+        if self.validate_func is not None:
+            if not self.validate_func(value):
+                raise ConfigException('Failed to validate {!r} config option'.format(self.name))
+        return value
+
+    def get(self):
+        value = self.store.get(self.name, self.default, self.section)
+        if self.cast is not None:
+            return self.cast(value)
+        return value
+
+    def set(self, value):
+        if self.cast is not None:
+            value = self.cast(value)
+        self.validate(value)
+        self.store.set(self.name, value, self.section)
+
+    @property
+    def value(self):
+        return self.get()
```

Comment:
why have this twice?

Reply:
Do we want to go with `option.get()` syntax or `option.value`?

Reply:
I like both, but we should really decide. `.get()` might lead some "clever" script writers to "optimize" the code by only fetching it once though, thus breaking config reloading.

Reply:
That brings up another question I had: do we want to go with the flow of getting each config value direct every time it's used (so that when the config is updated, it's automatically updated everywhere it's used), or work in a command or hooks system to call reload functions when we want to reload/change the config?

One problem I can see with automatically updating config values everywhere is that in fragile parts in the server internals it might not like it if the config value changes suddenly...

Reply:
yes, that's a concern, but those are just things we will have to iron out I'm afraid.

Luckily, the number of places the config would be changed from is pretty limited and since there is not a lot of concurrency, the parts of code that could be hit with this are limited too. The only thing that could be dangerous is with timed callbacks using `reactor.callLater()`

Reply:
I think I prefer `.get()` as it makes it more obvious that you are calling a method that is providing a value (which may not always be the same). With `.value` it feels more like a value that has no reason to change. Also we don't want anyone trying `option.value = 'something'`